### PR TITLE
[DNM]Fastmos III.5: Rogue Hashlists

### PR DESCRIPTION
**Pull request by Cyberboss**
TODO:

- [x] Fix the for loops @RemieRichards gave me the better way
- [x] Revert map edits and make a translator for filter_type
- [x] Fix shitty line ending changes
- [x] tgui needs a recompile, someone tell me how!
- [x] Make gas_mixtures pooled
- [x] Make excited_groups pooled
- [x] Sanic the for loops
- [x] Needs testmerging

**TL;DR I can't tell if it's better or worse, send help**

OK, essay time

For whatever reason, I lost my mental facilities for a moment and looked at LINDA and saw that /datum/gas_mixture used associative lists.

So, after recovering from the resulting alcoholism, I took it up and changed the entire thing to numbered lists. Basic computer science tells us that they are miles faster than the probable hashing that associative lists use. I'm bad at explaining, see @optimumtact's comment https://github.com/tgstation/tgstation/pull/22287#issuecomment-267797943 for the TL;DR

So, let's get down to syntax.

Before you'd access gasses like this:
```
var/datum/gas_mixture/gm = ...
var/oxy = gm.gases["o2"]
var/plasma = gm.gases["plasma"]
```
Now you do it like this
```
var/datum/gas_mixture/gm = ...
var/oxy = gm.gases[GAS_O2]
var/plasma = gm.gases[GAS_PLASMA]
```

Looping through contained gases has changed too, because `in` doesn't work in non-associative lists
Old:
```
var/cached_gases = gm.gases
for(var/id in cached_gases)
    ... = cached_gases[id][MOLES]
```
New:
```
var/cached_gases = gm.gases
for(var/gas in GAS_FOR(cached_gases))
   ... = gas[MOLES]
```
If you need access to the gas id use:
```
var/cached_gases = gm.gases
for(var/id in 1 to GAS_LAST)
    if(!cached_gases[id])  //entries may be null
        continue
   ... = cached_gases[id][MOLES]
```

Gas strings e.g. `"o2=22;TEMP=20"` are unaffected

Gas mixtures are now created using PoolOrNew

Also added procs for converting between the old and new ids `proc/gasid2shorthand` and `proc/shorthand2gasid` respectively. Also `proc/is_valid_gasid` for checking if an id is valid

Other changes:
- Qdeling and pooling of excited_groups and gas_mixtures
- Removed the strong dependencies on the old 4 hard_coded gases (o2,pl,co2,n2)
- Removed chrono_field/assume_air because the only this it did different than the base was leak the gas_mixture
- Fixed tank/blob_act double qdeling the tank
- Fixed an incorrectly set gas filter in pubbystation
- I noticed that turf/return_air and chrono_field/return_air leak GMs but I was unsure of how to handle this

Gonna need test merging. I'm 99% sure I got all the runtimes, but just in case.
I'll be watching merged PRs to keep this up to date

:cl: Cyberboss
experiment: Fastmos III.5 is here! Enjoy your faster fires and harder spacewinds!
/:cl:

Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding:

No

### CODE REVIEW

**Comment by MrPerson** [comment]
Would like to see a comparison profile of some kind of test case. I'm aware this would be hard to set up though

**Comment by Cyberboss** [comment]
Yeah, I'll work on that if I can figure out how

**Comment by ChangelingRain** [comment]
That doesn't look like a valid reason! Please provide a valid reason for: "Making atmos faster."
Clippy can help!

**Comment by Cyberboss** [comment]
@MrPerson Maybe The firesim + a maxcap?

**Comment by RemieRichards** [comment]
>because in doesn't work in non-associative lists

But this is incorrect?
So either you've explained yourself poorly or you're incorrect.

**Comment by optimumtact** [comment]
>So, after recovering from the resulting alcoholism, I took it up and changed the entire thing to keyed lists. >Basic computer science tells us that they are miles faster than the probable hashing that associative lists use.

Can you explain this further? I am not sure what you mean, byond only has one type of dictionary structure, the associative list. A keyed list and an associative list are the same thing?

As for the change from string -> number, yes, absolutely.

**Comment by Cyberboss** [comment]
@RemieRichards I tested it, it actually doesn't work. I was surprised too

**Comment by RemieRichards** [comment]
@Cyberboss You clearly didn't because I just did and it works fine, so back to:
* you either explained yourself poorly
* or you're incorrect

**Comment by optimumtact** [comment]
@RemieRichards why would you translate at runtime? Better to just use the defines he has added helpfully

**Comment by RemieRichards** [comment]
@optimumtact for the maps you twonk, he's saying the mappers have to right the indices now, instead of the "n2" etc. that's not going to help them when most of them hate code.

**Comment by Cyberboss** [comment]
It's not a big deal to translate them If you guys want

**Comment by RemieRichards** [comment]
I know it's not a big deal that's why I requested it, anything to not scare the very few mappers we have away.

**Comment by optimumtact** [comment]
okay I chatted with @Cyberboss in irc and all sorted now.

File: code/__HELPERS/unsorted.dm [hunk 1]
```diff
@@ -875,10 +875,11 @@ var/list/WALLITEMS_INVERSE = typecacheof(list(
 
 		var/list/cached_gases = air_contents.gases
 
-		for(var/id in cached_gases)
-			var/gas_concentration = cached_gases[id][MOLES]/total_moles
-			if(id in hardcoded_gases || gas_concentration > 0.001) //ensures the four primary gases are always shown.
-				user << "<span class='notice'>[cached_gases[id][GAS_META][META_GAS_NAME]]: [round(gas_concentration*100, 0.01)] %</span>"
+		for(var/id = 1 to GAS_LAST)
```
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: This is pedantic as hell, but we do this format as ```for(var/id in X to Y)```, not ```for(var/i = X to Y)```, there's no (major) difference in performance between these two, but we prefer the former, so I'd prefer you swap to those, so people don't go "hurr is there a new format for loops?"

Comment (Cyberboss) [inline code/__HELPERS/unsorted.dm]: K

Comment (optimumtact) [inline code/__HELPERS/unsorted.dm]: I thought the reason we preferred it was that it was faster for some reason

Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: @optimumtact it's faster than ```for(var/i = X i < Y; X++)```, both the above are, but they're equivalent to each other (near enough)

so there *are* 3 formats, 2 are fast and one's dogshit, we don't use the dogshit one.

**Comment by Cyberboss** [comment]
@RemieRichards Check this out I replaced the id thing in total_moles (which is used by breathing code) with `in` and had it generate a stack_trace and this is what happens
untitled

**Comment by RemieRichards** [comment]
@Cyberboss that's because you're putting the object back into itself.
an ```in``` loop gets you the object (unless keys, if associative/dict), not the indices, so you're getting an object out of ```cached_gasses``` and then trying to use it as a key (god knows why it passes the if though)

**Comment by Cyberboss** [comment]
Hmm I'll try it out then. Also it passes the if because ANY var thats not null or a number can be used as an associated key

**Comment by RemieRichards** [comment]
@Cyberboss yes, but for it to pass the if there has to be something AT that key, ```if(list["chickennuggets"])``` is only true if there's something *AT* the key ```"chickennuggets"```, since you're putting the object you just got out back in as a key, unless I'm very mistaken about listmos, there shouldn't be anything keyed to that object.

**Comment by Cyberboss** [comment]
It doesn't actually pass though, which is why there is no runtime and my dude is suffocating

**Comment by RemieRichards** [comment]
Oh right fair enough then I am correct about you misusing the object instead of index.
I *WAS* confused why you had the DD window up with nothing in it... forgot about stack_trace.

So tl;dr it was just you being silly thinking ```in``` got you an index :P

**Comment by Cyberboss** [comment]
@RemieRichards How about we both blame BYOND and call it a day

**Comment by RemieRichards** [comment]
Well no real language would provide you the indices when looping through it's arrays or lists (unless they're dicts) so technically byond *is* in the right here, but given how weird byond is I can say it's more likely you assumed it to be insane and worked around that :P

File: code/__HELPERS/unsorted.dm (hunk 1 above)
Comment (MrPerson) [inline code/__HELPERS/unsorted.dm]: They're actually identical afaik. Not that that changes your request or anything

Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: @MrPerson I have seen *VERY* minor differences between = X to Y and in X to Y, we're talking 0.0001s of difference, far less than anything actually worth acting on though, yes.

**Comment by karlnp** [comment]
you should be able to just run `gulp --min` or `build_assets.bat` in the tgui directory and it'll build everything. i mean it'll vomit garbage but the end product will work anyway. javascript

**Comment by xxalpha** [comment]
Wasn't it numbered before dunc changed it to assoc lists?

**Comment by optimumtact** [comment]
```
## WARNING: Wrong gas ID in the gas filter's filter_type var. filter_type == -1 in code/modules/atmospherics/machinery/components/trinary_devices/filter.dm at line 90 src: the gas filter usr: .
## WARNING: Wrong gas ID in the gas filter's filter_type var. filter_type == -1 in code/modules/atmospherics/machinery/components/trinary_devices/filter.dm at line 90 src: the gas filter usr: .
## WARNING: Wrong gas ID in the gas filter's filter_type var. filter_type == -1 in code/modules/atmospherics/machinery/components/trinary_devices/filter.dm at line 90 src: the gas filter usr: .
```

Are these runtimes from basil related to this pr, or are they just general breakage.

**Comment by Cyberboss** [comment]
Not testmerged, not me

**Comment by optimumtact** [comment]
Can you fix conflicts and I'll go beat the admins to get this test merged

**Comment by Cyberboss** [comment]
Done, don't be in a rush to fully merge this. I wanna see if it's actually even worth it because of what @xxalpha said

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 2]
```diff
@@ -445,17 +490,26 @@ var/list/gaslist_cache = null
...
+	var/unique_sent_gases = FALSE
 
 	//GAS TRANSFER
-	for(var/id in sharer_gases - cached_gases) // create gases not in our cache
-		add_gas(id)
-	for(var/id in cached_gases) // transfer gases
-		if(!sharer_gases[id]) //checking here prevents an uneeded proc call if the check fails.
-			sharer.add_gas(id)
-
+	for(var/id in 1 to GAS_LAST) // create gases not in our cache
```
Comment (Cyberboss) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Potentially the most unperformant part of this PR (this is proc/share).

It needs the gasid which you can't get with fastfors. Discuss

Comment (MrStonedOne) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Store the ID in the gas list (most likely in meta)
then you can go back to sharer_gases - cached_gases

**Comment by MrStonedOne** [comment]
@duncathan you might be interested in this

**Comment by MrStonedOne** [comment]
do not merge, I see some things that might be an issue, also could be causing #22456

File: code/controllers/subsystem/air.dm [hunk 3]
```diff
@@ -268,7 +268,7 @@ var/datum/subsystem/air/SSair
 		active_turfs -= T
 		if (T.blocks_air)
 			continue
-		T.Initalize_Atmos(times_fired)
+		T.Initialize_Atmos(times_fired)
```
Comment (MrStonedOne) [inline code/controllers/subsystem/air.dm]: It really makes me sad to see you include unrelated changes like this in your prs.

Today mostly because my testing/profiling setup requires I edit this file, and editing this file 4 or 5 times is lame and it would have been nice if i could have just copy and pasted my changes in as one block, but now i can't, and its not even for a good reason.

You fuck yourself over with conflicts, you fuck other people over with conflicts, and its really not necessary.

File: code/game/turfs/space/space.dm [hunk 4]
```diff
@@ -12,8 +12,8 @@
 	var/destination_x
 	var/destination_y
 
-	var/global/datum/gas_mixture/space/space_gas = new
-	plane = PLANE_SPACE
+	var/global/datum/gas_mixture/space/space_gas = init_space_and_gaslists()
+	plane = PLANE_SPACE
```
Comment (MrStonedOne) [inline code/game/turfs/space/space.dm]: What purpose does turning this new() into a factory provide?

File: code/modules/atmospherics/gasmixtures/space_mixture.dm [hunk 5]
```diff
@@ -1,5 +1,7 @@
 //"immutable" gas mixture used for space calculations
 //it can be changed, but any changes will ultimately be undone before they can have any effect
+#define SPACE_MIX_HEAT_CAP 7000
+#define GET_SPACE_MIX PoolOrNew(/datum/gas_mixture/space)
```
Comment (MrStonedOne) [inline code/modules/atmospherics/gasmixtures/space_mixture.dm]: pool or new for a singleton is fucking retarded.

File: code/game/turfs/simulated/floor/fancy_floor.dm [hunk 6]
```diff
@@ -38,8 +38,9 @@
 
 /turf/open/floor/grass/New()
 	..()
-	spawn(1)
-		update_icon()
+
+/turf/open/floor/grass/initialize()
+	update_icon()
```
Comment (MrStonedOne) [inline code/game/turfs/simulated/floor/fancy_floor.dm]: unrelated change

File: code/game/turfs/simulated/floor/fancy_floor.dm [hunk 7]
```diff
@@ -98,8 +99,9 @@
 
 /turf/open/floor/carpet/New()
 	..()
-	spawn(1)
-		update_icon()
+
+/turf/open/floor/carpet/initialize()
+	update_icon()
```
Comment (MrStonedOne) [inline code/game/turfs/simulated/floor/fancy_floor.dm]: unrelated change

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 8]
```diff
@@ -6,31 +6,53 @@ What are the archived variables for?
...
-	else
-		cached_gas = gaslist_cache[id]
-	//Copy() it because only GAS_META is static
+#define INIT_GASES new(gaslist_cache.len)	//To be used to setup an empty gases var
+
+var/list/meta_gas_info	//contains the static information of a gas type
+
+var/list/gaslist_cache	//pre initialized list of gases keyed by id, ONLY FOR COPYING
+
+var/list/null_gases	//for sanic loops
```
Comment (MrStonedOne) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: world already has an init() proc.

you should be using factories on these, not defining them on first use.

You aren't saving any overhead, and actually creating overhead by having to make sure these lists exist.

In the land of atmos, even a unneeded if() is a sin.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 9]
```diff
@@ -42,13 +64,19 @@ var/list/gaslist_cache = null
...
 	temperature = 0
 	temperature_archived = 0
 	src.volume = volume
 	last_share = 0
 	fuel_burnt = 0
 
+/datum/gas_mixture/Destroy()
+	gases = null
+	holder = null
+	. = QDEL_HINT_PUTINPOOL
```
Comment (MrStonedOne) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: I'm gonna tell you right now that the pooling is gonna be likely to be causing an overhead that wouldn't exist without it, both on qdel, and on fetching a item out of the pool.

byond already does pooling internally, and the cost of creating new objects is in the New(), not in creating the actual object, something the pooling doesn't properly address.

I've wanted to believe, but pooling is currently more costly because it has to call New() still, and object creation overhead is rarely above just the overhead of the proc call to poolornew.

This change will most likely have to go. I'll see about profiling what this change added if it's easy to revert it


As an aside, atmos doesn't use qdel because it's a closed system, we don't do checks in atmos when we don't have to, this includes checking for objects leaking (all that qdel basically does) this overhead is likely to not be worth it.

atm a gas_mixture should never be needed to be explicitly deleted, if something is referencing it, it needs to exist, so it's better to just let byond's GC handling work this out and save the overhead in temp gas mixtures. 
If a gas mixture is ever needed to be hard deleted, it is more likely the bug is in the fact it was ever sent to qdel(ie something was being too eager to qdel references on it), an issue that doesn't exist if you stick to byond's built-in referencing counting stuff.

**Comment by Cyberboss** [comment]
Blah, you go first, I'll try fixing this shit sometime this week

**Comment by RemieRichards** [comment]
>sometime this week
>christmas day

wew.

**Comment by Cyberboss** [comment]
Ill fix Atmos before bringing this back

**Comment by duncathan** [comment]
(mega necro but idc)

btw this implementation causes significant performance hits in the average case (a roughly default air mixture in a tile) for each new gas added and that is why i chose to implement it as an assoc list. i profiled the average case by running identical plasma fires (meaning each tile would have a mixture with the 4 default gases in it) throughout a default station. with the number of gases we had when I wrote listmos, assoc lists were slower. however, with some relatively close number `n` of gases defined (it was on the order of 12-16, where we had 7 gases already defined) assoc lists were faster. of course, assoc lists are much worse in the worst case (all gases present in a mixture); however, those cases are extremely rare and get rarer and rarer as we add more gases. with assmos approaching a merge-ready state rapidly, and more gases yet to come, it's essentially a non-question as to whether we should continue using assoc lists or attempt to switch to indexed lists.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 9 above)
Comment (optimumtact) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm:77]: you were right
